### PR TITLE
[Enhancement] Add GroupByCountDistinctRewriteRule

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Rewrite count(distinct xx) group by x when distinct columns satisfy scan distribution
e.g.
select count(distinct xx) from t group by x
->
select count(xx) from (select x, xx from t group by x, xx) tt group by x;
 
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java [hunk 1]
```diff
@@ -274,8 +274,8 @@ public class RuleSet {
         ));
 
         rewriteRules.put(RuleSetType.MULTI_DISTINCT_REWRITE, ImmutableList.of(
-                new RewriteMultiDistinctRule(),
-                new RewriteMultiDistinctByCTERule()
+                new RewriteMultiDistinctByCTERule(),
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java:277]: Why change the order?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 100 / 100 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|91|91|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggregate = (LogicalAggregationOperator) input.getOp();
+        LogicalOlapScanOperator scan = (LogicalOlapScanOperator) input.getInputs().get(0).getOp();
+
+        Collection<CallOperator> calls = aggregate.getAggregations().values();
+
+        List<CallOperator> distinctList = calls.stream().filter(this::isDistinct).collect(Collectors.toList());
+        List<CallOperator> otherList = calls.stream().filter(c -> !isDistinct(c)).collect(Collectors.toList());
+
+        // check other function
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:86]: add some comment for why we only support these function

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                .map(c -> (ColumnRefOperator) c).distinct().collect(Collectors.toList());
+        if (distinctColumns.size() != 1) {
+            return false;
+        }
+
+        // check distribution satisfy scan node
+        List<Integer> groupBy = aggregate.getGroupingKeys().stream().map(ColumnRefOperator::getId)
+                .collect(Collectors.toList());
+
+        if (groupBy.isEmpty() || groupBy.containsAll(scan.getDistributionSpec().getShuffleColumns())) {
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:117]: Why return false when groupBy columns satisfied scan's shuffle columns?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                    getAggregation(secondFn, new Type[] {firstAgg.getType()}, Lists.newArrayList(firstOutput));
+            secondAggregations.put(k, secondAgg);
+        });
+
+        distinctMap.forEach((k, v) -> {
+            Function origin = v.getFunction();
+            String secondFn = DISTINCT_FUNCTION_TRANS.get(origin.getFunctionName().getFunction());
+            CallOperator secondAgg = getAggregation(secondFn, origin.getArgs(), v.getChildren());
+            secondAggregations.put(k, secondAgg);
+        });
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:173]: Distinct only has second aggregation?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java:277]: rewrit to cte first, and rewrite other to multi distinct if cte not support, don't depend on the cte flag

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java (hunk 3 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:117]: don't need rewrite by this rule, origin plan may better

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java (hunk 4 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:173]: For count distinct column, first stage: group by(distinct), second stage: count the column... you can see I put the distinct column to GroupBy in first stage

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 100 / 100 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|91|91|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            .put(FunctionSet.COUNT, FunctionSet.COUNT)
+            .put(FunctionSet.SUM, FunctionSet.SUM)
+            .put(FunctionSet.MULTI_DISTINCT_SUM, FunctionSet.SUM)
+            .put(FunctionSet.MULTI_DISTINCT_COUNT, FunctionSet.COUNT)
+            .build();
+
+    // multi-stage other function mapping
+    // origin aggregate -> Pair<1st stage, 2nd stage>
+    // these aggregate function can be support multi-stage
+    private static final Map<String, Pair<String, String>> OTHER_FUNCTION_TRANS =
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:57]: Why need this map? keep DISTINCT_FUNCTION_TRANS isn't enough?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+                    .put(FunctionSet.HLL_UNION, Pair.create(FunctionSet.HLL_UNION, FunctionSet.HLL_UNION))
+                    .put(FunctionSet.NDV, Pair.create(FunctionSet.HLL_UNION, FunctionSet.NDV))
+                    .put(FunctionSet.BITMAP_UNION, Pair.create(FunctionSet.BITMAP_UNION, FunctionSet.BITMAP_UNION))
+                    .put(FunctionSet.BITMAP_UNION_COUNT,
+                            Pair.create(FunctionSet.BITMAP_UNION, FunctionSet.BITMAP_UNION_COUNT))
+                    .put(FunctionSet.PERCENTILE_UNION,
+                            Pair.create(FunctionSet.PERCENTILE_UNION, FunctionSet.PERCENTILE_UNION))
+                    .build();
+
+    public GroupByCountDistinctRewriteRule() {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:72]: What if there is project above scan operator?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java [hunk 7]
```diff
@@ -169,6 +170,10 @@ public int hashCode() {
 
         @Override
         public LogicalAggregationOperator build() {
+            Preconditions.checkNotNull(type);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java:173]: What's the meaning of these checks?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        LogicalAggregationOperator first =
+                new LogicalAggregationOperator(AggType.GLOBAL, firstGroupBy, firstAggregations);
+        second.withOperator(aggregate).setAggregations(secondAggregations);
+
+        return Lists.newArrayList(
+                OptExpression.create(second.build(), OptExpression.create(first, input.getInputs())));
+    }
+
+    @NotNull
+    private CallOperator getAggregation(String name, Type[] argTypes, List<ScalarOperator> args) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:185]:
```suggestion
    private CallOperator genAggregation(String name, Type[] argTypes, List<ScalarOperator> args) {
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java [hunk 9]
```diff
@@ -48,11 +48,6 @@ public RewriteMultiDistinctRule() {
     @Override
     public boolean check(OptExpression input, OptimizerContext context) {
         LogicalAggregationOperator agg = (LogicalAggregationOperator) input.getOp();
-        boolean hasNoGroup = agg.getGroupingKeys().isEmpty();
-        // check cbo is enabled and hasNoGroup is true
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java:52]: why remove this?

**Comment by kangkaisen** [comment]
Please add the test case the benchmark framework.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java (hunk 9 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java:52]: unused

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java (hunk 7 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java:173]: These are all necessary paramters

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java (hunk 6 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:72]: project has been merged

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java (hunk 5 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:57]: DISTINCT_FUNCTION_TRANS don't need first stage

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 100 / 100 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|91|91|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 102 / 102 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|93|93|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Review by Youngwb — DISMISSED** [review]

**Review by HangyuanLiu — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 102 / 102 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|93|93|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Review by HangyuanLiu — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**